### PR TITLE
Live blog contributor images

### PR DESCRIPTION
**Pull request by JustinPinner**
Display contributor's byline image and name on liveblog posts if available, and if the block is properly marked with data-block-contributor, and if the feature is switched on.

Styling is the same as was used for the liveblog coverage of the oscars (http://www.theguardian.com/film/filmblog/live/2015/feb/22/oscars-2015-live-red-carpet-and-arrivals-ceremony-and-winners).

### CODE REVIEW

File: common/app/views/support/HtmlCleaner.scala [hunk 1]
```diff
@@ -292,6 +292,21 @@ case class LiveBlogDateFormatter(isLiveBlog: Boolean)(implicit val request: Requ
   }
 }
 
+case class BloggerBylineImage(article: Article)(implicit val request: RequestHeader) extends HtmlCleaner  {
+  def clean(body: Document): Document = {
+    if (article.isLiveBlog && LiveBlogContributorImagesSwitch.isSwitchedOn) {
+      body.select(".block").foreach { el =>
+        val contributorId = el.attributes().get("data-block-contributor")
+        article.tags.find(_.id == contributorId).map{ contributorTag =>
+          val html = views.html.fragments.meta.bylineLiveBlockImage(contributorTag)
```
Comment (gklopper) [inline common/app/views/support/HtmlCleaner.scala]: Like this use of a template

**Comment by gklopper** [comment]
Got some screenshots?

File: common/app/conf/switches.scala [hunk 2]
```diff
@@ -228,6 +228,12 @@ object Switches {
     safeState = On, sellByDate = never
   )
 
+  val LiveBlogContributorImagesSwitch = Switch("Feature", "liveblog-contributor-image",
+    "Show contributor byline images in live blog blocks",
+    safeState = On,
```
Comment (gklopper) [inline common/app/conf/switches.scala]: You mean for this to be `On`?

File: common/app/views/support/HtmlCleaner.scala [hunk 3]
```diff
@@ -292,6 +292,21 @@ case class LiveBlogDateFormatter(isLiveBlog: Boolean)(implicit val request: Requ
   }
 }
 
+case class BloggerBylineImage(article: Article)(implicit val request: RequestHeader) extends HtmlCleaner  {
+  def clean(body: Document): Document = {
+    if (article.isLiveBlog && LiveBlogContributorImagesSwitch.isSwitchedOn) {
+      body.select(".block").foreach { el =>
+        val contributorId = el.attributes().get("data-block-contributor")
+        article.tags.find(_.id == contributorId).map{ contributorTag =>
```
Comment (gklopper) [inline common/app/views/support/HtmlCleaner.scala]: Does this assume every block has a `data-block-contributor`?

Should you rather...

    Option(el.attributes().get("data-block-contributor")).foreach{ id =>
      ......
    }

**Comment by gklopper** [comment]
:+1:

File: common/app/conf/switches.scala (hunk 2 above)
Comment (JustinPinner) [inline common/app/conf/switches.scala]: I considered defaulting to an `Off` state but thought that'd disable the feature after every deploy. If that's not the case I'll switch it around. No pun intended.

File: common/app/views/support/HtmlCleaner.scala (hunk 3 above)
Comment (JustinPinner) [inline common/app/views/support/HtmlCleaner.scala]: Oops, no I can't rely on every block of every live blog having the attribute. I like your suggestion better than relying on the .tags.find failing implicitly so will change that.

Comment (gklopper) [inline common/app/views/support/HtmlCleaner.scala]: Just check my solution actually works :smile: it came straight out of my head and never went through a compiler. Also I am assuming that `el.attributes().get("data-block-contributor")` returns `null` if there is no attribute (it is a Java library not a Scala one).

Comment (gklopper) [inline common/app/views/support/HtmlCleaner.scala]: You should put some tests in `ArticleFeatureTest` around these.

**Comment by JustinPinner** [comment]
There's a single screenshot available (currently the only article we have for testing I think);
screen shot 2015-04-14 at 16 06 48

File: common/app/views/support/HtmlCleaner.scala (hunk 3 above)
Comment (JustinPinner) [inline common/app/views/support/HtmlCleaner.scala]: It looks like null is converted to an empty string in attributes.get. So maybe;
`val contributorId = el.attributes().get("data-block-contributor")`
`if (contributorId.nonEmpty) {`
`  ...` 
`}`

Comment (gklopper) [inline common/app/views/support/HtmlCleaner.scala]: As long there are no nulls involved your original code is probably fine. It basically says `if (profileId == "")`

File: common/app/conf/switches.scala (hunk 2 above)
Comment (robertberry) [inline common/app/conf/switches.scala]: Probably should be `Off` to begin with. Then when this expires, we'll switch it to `On` and set the sellByDate to never.

I think the boxes tend to load the switches pretty sharpish when they start up so you probably wouldn't see this often (if at all).

File: static/src/stylesheets/module/content/_live-blog.scss [hunk 4]
```diff
@@ -271,6 +271,27 @@ $block-padding-right: $gs-gutter;
 
     .block-elements {
 
+        /* Bylines
+           ========================================================= */
+        .liveblog-block-byline {
+            border-bottom: 1px dotted #dcdcdc;
+            padding-bottom: 8px;
```
Comment (robertberry) [inline static/src/stylesheets/module/content/_live-blog.scss]: Should generally do vertical margins as calculations on $gs-baseline and horizontal margins as calculations on $gs-gutter

See https://github.com/guardian/guss-grid-system

Comment (robertberry) [inline static/src/stylesheets/module/content/_live-blog.scss]: Actually generally get @sndrs to check this over. He's better at this stuff than I am :-) But things like colours here should be using variables, and I think we have mixins for fonts.

**Comment by robertberry** [comment]
:+1: Nice work

File: static/src/stylesheets/module/content/_live-blog.scss (hunk 4 above)
Comment (JustinPinner) [inline static/src/stylesheets/module/content/_live-blog.scss]: Ok, thanks. Will investigate how to do that.

Comment (sndrs) [inline static/src/stylesheets/module/content/_live-blog.scss]: Yeah @robertberry is right - we use those two variables in Sass to try and keep consistent spacing. 

The variables are set in this bower package: https://github.com/guardian/guss-grid-system/blob/master/_grid-system.scss (but tl;dr, `$gs-baseline` is `16px`, `$gs-gutter` is `20px`).

So if you need 8 vertical pixels, you would use `$gs-baseline * (2/3)`. It's worth noting though that if you're not just halving, doubling etc, it's a bit of a code/design smell (though obviously not always, if you need 8 then you need 8).

File: static/src/stylesheets/module/content/_live-blog.scss [hunk 5]
```diff
@@ -271,6 +271,27 @@ $block-padding-right: $gs-gutter;
...
+        }
+        .liveblog-block-byline__img {
+            display: inline;
+            border-radius: 50%;
+            max-width: 36px;
+            vertical-align: middle;
+            margin-right: 4px;
```
Comment (sndrs) [inline static/src/stylesheets/module/content/_live-blog.scss]: `$gs-gutter * .2`

Comment (JustinPinner) [inline static/src/stylesheets/module/content/_live-blog.scss]: Thanks Alex. I'll drop those in and test.



On 16 April 2015 at 11:06, Alex Sanders <notifications@github.com> wrote:

> In static/src/stylesheets/module/content/_live-blog.scss
> <https://github.com/guardian/frontend/pull/8865#discussion_r28497648>:
>
> > @@ -271,6 +271,27 @@ $block-padding-right: $gs-gutter;
> >
> >      .block-elements {
> >
> > +        /* Bylines
> > +           ========================================================= */
> > +        .liveblog-block-byline {
> > +            border-bottom: 1px dotted #dcdcdc;
> > +            padding-bottom: 8px;
> > +        }
> > +        .liveblog-block-byline__img {
> > +            display: inline;
> > +            border-radius: 50%;
> > +            max-width: 36px;
> > +            vertical-align: middle;
> > +            margin-right: 4px;
>
> $gs-gutter * .2
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/8865/files#r28497648>.
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

File: static/src/stylesheets/module/content/_live-blog.scss [hunk 6]
```diff
@@ -271,6 +271,27 @@ $block-padding-right: $gs-gutter;
...
+        }
+        .liveblog-block-byline__name {
+            font-family: 'EgyptianText', Georgia;
+            font-weight: normal;
+            font-size: 16px;
```
Comment (sndrs) [inline static/src/stylesheets/module/content/_live-blog.scss]: Similar to lengths, type is defined by a bunch of presets in https://github.com/guardian/guss-typography. there's a graphic demoing the sizes/look etc at https://raw.githubusercontent.com/guardian/guss-typography/master/font-scale.png, but for this, you'd use `@include fs-bodyCopy(2)`.

File: static/src/stylesheets/module/content/_live-blog.scss [hunk 7]
```diff
@@ -271,6 +271,27 @@ $block-padding-right: $gs-gutter;
...
+            display: inline;
+            color: #333333;
```
Comment (sndrs) [inline static/src/stylesheets/module/content/_live-blog.scss]: Colours get the same treatment, at https://github.com/guardian/pasteup-palette. This is more about chosing a colour by intention or variant thereof, but if you know you need that specific value, `$neutral-1` maps to that (demo here http://guardian.github.io/pasteup-palette/demo/).

File: static/src/stylesheets/module/content/_live-blog.scss [hunk 8]
```diff
@@ -271,6 +271,25 @@ $block-padding-right: $gs-gutter;
 
     .block-elements {
 
+        /* Bylines
+           ========================================================= */
+        .liveblog-block-byline {
+            border-bottom: 1px dotted #dcdcdc;
```
Comment (sndrs) [inline static/src/stylesheets/module/content/_live-blog.scss]: `neutral-4` - sorry, missed this one earlier. basically, the only colour you should spec manually is `#ffffff`

**Comment by sndrs** [comment]
:+1: from me too, apart from last comment

File: static/src/stylesheets/module/content/_live-blog.scss (hunk 8 above)
Comment (JustinPinner) [inline static/src/stylesheets/module/content/_live-blog.scss]: Doh! I should've questioned that. Ok, will fix.

**Pull request merged**